xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-api/src/main/java/org/xwiki/panels/internal/AbstractPanelsUIExtensionManager.java

```diff
@@ -126,8 +130,12 @@
                         extensionId = resolver.resolve(extension.getId());
                     }
 
-                    if (panelSerializedReferences.contains(extensionId)) {
-                        panels.add(extension);
+                    if (panelReferenceWithPosition.containsKey(extensionId)) {
+                        int position = panelReferenceWithPosition.get(extensionId);
+                        if (position > panels.size()) {
+                            position = panels.size();
+                        }
+                        panels.add(position, extension);
```

Comment:
This looks complex. Can't you achieve the same with a LinkedSet instead of a HashSet?

Reply:
Also if I remember we were iterating on the panelSerializedReferences before, need to check

Reply:
No because I don't iterate on the panels element but on the UIExtensions we get, so I don't know in which order I'm building the new list

Reply:
Yes we had this which was guaranteing the order:

```
for (String panelSerializedReference : panelSerializedReferences) {
```

Reply:
So why not put that back?

Reply:
Yes I changed that to avoid iterating multiple times on the UIExtension, as the list might be big.

Reply:
indeed, you improved the perfs, that's good!

Reply:
Another idea: why not use `indexOf()` instead of `contains()` so that you get both the index + whether it exists, to replace: `if (panelSerializedReferences.contains(extensionId)) {` ?

It's slightly less efficient though than your solution, at the expense of less memory.

Reply:
I personally find my solution more clear, as it emphasizes the fact that you won't iterate on the panel references. But it's really more a matter of style there.

Reply:
My preference goes to conciseness in general. But I'm fine to keep your solution.